cmd/cellrangerBam/cellrangerBam.go

```diff
@@ -0,0 +1,823 @@
+			if !found {
+				mp[columns[1]] = 1
+			} else {
+				mp[columns[1]] = count + 1
+			}
+		}
+		if norm != "" {
+			inputNormalize(mp, norm)
+		}
+		for i = range mp {
+			toWrite = append(toWrite, fmt.Sprintf("%s\t%f\t%c", i, mp[i], whichBin))
+		}
+		sort.Strings(toWrite)
+		for _, i = range toWrite {
+			fileio.WriteToFileHandle(out, i)
+		}
+		whichBin++
+	}
+}
+
+// determine ideal bins will determine how many bins should be used for the pseudobulk -binCells option
+func determineIdealBins(s Settings, cellTypeSlice []string) int {
+	var count, j int
+	var found, stop bool
+	var z, l, m string
+	var binSlice []int
+	var matrix [][]string
+
+	nc := fileio.Read(s.determineBins)
+
+	for i := 0; i <= 10; i++ { //do this whole loop 10 times (essentially 10 replicates)
+		stop = false
+		for j = 1; j <= 100; j++ { //loop from bins 1 to 100 (j)
+			s.binCells = j                                   //set the binCells parameter to j
+			matrix = distributeCells(s, cellTypeSlice, true) //run distributeCells with that value of j
+			for k := range matrix {                          //loop through bins in matrix
+				ncMap := make(map[string]int) //create a map that has ncConstructName--counts
+				for _, z = range nc {         //populate the map
+					ncMap[z] = 0
+				}
+				for _, l = range matrix[k] { //loop through the constructs in a particular bin
+					columns := strings.Split(l, "\t") //split an entry in from the matrix
+					count, found = ncMap[columns[1]]  //search the 2nd column to see if it is a negative control
+					if found {                        //if it is, add a count to the map for that particular negative control
+						ncMap[columns[1]] = count + 1
+					}
+				}
+				for m = range ncMap {
+					count, _ = ncMap[m]
+					if count < 1 {
+						binSlice = append(binSlice, j-1)
+						stop = true
+						break
+					}
+				}
+				if stop == true {
+					break
+				}
+			}
+			if stop == true {
+				break
+			}
+		}
+	}
+	return int(numbers.AverageInt(binSlice))
+}
+
+// determineBin takes in an int and the probablity of any given bin and determines the bin the cell belongs to.
+func determineBin(numBins int, prob float64) int {
+	var bin int
+
+	randNum := rand.Float64()      // draw a random number
+	for j := 0; j < numBins; j++ { //iterate over the number of bins
+		if j == numBins { //if we've gotten to the last bin and the cell still hasn't been partitioned, we don't have to check anything else we can just put it there.
+			return j
+		}
+		if randNum >= prob*float64(j) && randNum < prob*float64(j+1) { // check to see if our random number belongs to that bin
+			bin = j
+		}
+	}
+	return bin
+}
+
+// distributeCells takes in a slice of string that is created in parseBam that has a list of cellBarcodes and constructs. It will partition those cells and constructs into separate slices. The number of bins is a user-input variable
+func distributeCells(s Settings, cellTypeSlice []string, fromDB bool) [][]string {
```

Comment:
What does fromDB mean?

Reply:
that is "from determine bins". Normally the distributeCells function prints out how many cells make it into each bin, but i didn't want that to be printed out each time that determineBins runs the loop. So if fromDB is true, it doesn't print out the summary statistics for bin allocation